File: openmesh-apple/MeshFluxIos/core/GoEngine.swift
import Foundation
import OpenMeshGo

enum GoEngineError: LocalizedError {
        /// 核心库（OpenMeshGo）加载失败，无法使用钱包与 VPN 能力。
        case newLibReturnedNil
        
        var errorDescription: String? {
                switch self {
                case .newLibReturnedNil:
                        return "无法加载核心库，请重新安装应用或联系支持。"
                }
        }
}

final class GoEngine {
        static let shared = GoEngine()
        
        private let queue = DispatchQueue(label: "meshflux.go.engine.serial")
        
        private var lib: (any OpenmeshAppLibProtocol)?
        private var cachedConfig: Data = Data()
        
        private var initTask: Task<Void, Error>?

        private func maskAddress(_ address: String) -> String {
                let t = address.trimmingCharacters(in: .whitespacesAndNewlines)
                guard t.count > 12 else { return t }
                return "\(t.prefix(6))...\(t.suffix(4))"
        }

        private func log(_ message: String) {
                NSLog("GoEngine: %@", message)
        }
        
        /// 懒初始化：不在 init() 中创建 initTask，避免启动时阻塞导致 watchdog 超时
        private init() {
                self.cachedConfig = Data()
                // 不再在 init 中启动初始化任务
                // initTask 将在第一次调用 ensureReady() 时惰性创建
        }
        
        func generateMnemonic12() async throws -> String {
                try await ensureReady()
                
                return try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        let s = try lib.generateMnemonic12()
                                        cont.resume(returning: s)
                                } catch {
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
        
        /// ✅ 新增：调用 Go 的 CreateEvmWallet(mnemonic, pin) -> JSON string
        func createEvmWallet(mnemonic: String, pin: String) async throws -> String {
                try await ensureReady()
                
                return try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        let json = try lib.createEvmWallet(mnemonic, password: pin)
                                        cont.resume(returning: json)
                                } catch {
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
        
        /// ✅ 新增：调用 Go 的 DecryptEvmWallet(keystoreJSON, pin) -> WalletSecretsV1
        func decryptEvmWallet(keystoreJSON: String, pin: String) async throws -> WalletSecretsV1 {
                try await ensureReady()
                
                return try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        let secrets = try lib.decryptEvmWallet(keystoreJSON, password: pin)
                                        cont.resume(returning: secrets)
                                } catch {
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
        
        /// ✅ 新增：调用 Go 的 GetTokenBalance(address, tokenName, networkName) -> balance string
        func getTokenBalance(address: String, tokenName: String, networkName: String) async throws -> String {
                let requestID = UUID().uuidString.lowercased()
                log("getTokenBalance start request_id=\(requestID) address=\(maskAddress(address)) token=\(tokenName) network=\(networkName)")
                try await ensureReady()
                
                return try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        self.log("getTokenBalance invoking Go bridge request_id=\(requestID)")
                                        let balance = try lib.getTokenBalance(address, tokenName: tokenName, networkName: networkName)
                                        self.log("getTokenBalance success request_id=\(requestID) balance=\(balance)")
                                        cont.resume(returning: balance)
                                } catch {
                                        self.log("getTokenBalance failed request_id=\(requestID) error=\(String(describing: error))")
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
        
        /// ✅ 新增：调用 Go 的 GetSupportedNetworks() -> JSON string
        func getSupportedNetworks() async throws -> [String] {
                try await ensureReady()
                
                return try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        let networksJSON = try lib.getSupportedNetworks()
                                        guard let data = networksJSON.data(using: .utf8) else {
                                                throw NSError(domain: "GoEngine", code: -1, userInfo: [NSLocalizedDescriptionKey: "支持网络数据无效"])
                                        }
                                        
                                        let networks = try JSONDecoder().decode([String].self, from: data)
                                        cont.resume(returning: networks)
                                } catch {
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
        
        func reconfigure(config: Data) async throws {
                try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                self.cachedConfig = config
                                self.initTask = Task { [weak self] in
                                        guard let self else { return }
                                        try await self.initLocked(config: config)
                                }
                                cont.resume()
                        }
                }
                
                try await ensureReady()
        }
        
        func reset() async {
                await withCheckedContinuation { cont in
                        queue.async {
                                self.lib = nil
                                self.cachedConfig = Data()
                                self.initTask = nil
                                cont.resume()
                        }
                }
        }

        /// 仅释放 Go runtime 对象，保留已缓存配置，供前台快速恢复。
        /// 用于 iOS 退到后台后的内存回收，避免影响后续 Go 交互路径。
        func releaseRuntimeForMemoryPressure() async {
                log("releaseRuntimeForMemoryPressure begin")
                await withCheckedContinuation { cont in
                        queue.async {
                                self.lib = nil
                                self.initTask = nil
                                cont.resume()
                        }
                }
                log("releaseRuntimeForMemoryPressure end")
        }
        
        // MARK: - Internal
        private func ensureReady() async throws {
                if initTask == nil {
                        initTask = Task { [weak self] in
                                guard let self else { return }
                                try await self.initLocked(config: self.cachedConfig)
                        }
                }
                try await initTask?.value
        }
        
        private func initLocked(config: Data) async throws {
                try await withCheckedThrowingContinuation { cont in
                        queue.async {
                                do {
                                        self.cachedConfig = config
                                        
                                        if self.lib == nil {
                                                guard let real = OMOpenmeshNewLib() else {
                                                        throw GoEngineError.newLibReturnedNil
                                                }
                                                self.lib = OpenmeshAppLibBridge(omLib: real)
                                        }
                                        guard let lib = self.lib else {
                                                throw GoEngineError.newLibReturnedNil
                                        }
                                        
                                        try lib.initApp(config)
                                        
                                        cont.resume()
                                } catch {
                                        cont.resume(throwing: error)
                                }
                        }
                }
        }
}

// MARK: - VPN status (for HomeTabView etc.)
extension GoEngine {
        /// 返回当前 VPN 状态，供 UI 使用；若尚未就绪或为 Stub 则返回占位。
        func getVpnStatus() async -> VpnStatus? {
                do {
                        log("getVpnStatus start")
                        try await ensureReady()
                        return await withCheckedContinuation { cont in
                                queue.async {
                                        let status = self.lib?.getVpnStatus()
                                        if let status {
                                                self.log("getVpnStatus success connected=\(status.connected) server=\(status.server)")
                                        } else {
                                                self.log("getVpnStatus success nil")
                                        }
                                        cont.resume(returning: status)
                                }
                        }
                } catch {
                        log("getVpnStatus failed error=\(String(describing: error))")
                        return nil
                }
        }
}

// MARK: - Install bootstrap (debug)
extension GoEngine {
        private static let firstLaunchKey = "meshflux.didRunAfterInstall"
        
        /// 新安装后的第一次启动：清理 Keychain 残留（DEBUG 默认开启）
        static func bootstrapOnFirstLaunchAfterInstall() {
                let ud = UserDefaults.standard
                if ud.bool(forKey: firstLaunchKey) { return }
                ud.set(true, forKey: firstLaunchKey)
                
                try? PINStore.clear()
                try? WalletStore.clear() 
        }
}

File: openmesh-android/app/src/main/java/com/meshnetprotocol/android/core/GoEngine.kt
package com.meshnetprotocol.android.core

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Go 引擎错误类型
 */
sealed class GoEngineError(message: String) : Exception(message) {
    object NewLibReturnedNil : GoEngineError("无法加载核心库，请重新安装应用或联系支持。")
    object NotInitialized : GoEngineError("Go 引擎尚未初始化")
}

/**
 * Go 引擎单例
 * 管理 OpenMesh Go 库的生命周期和初始化
 * 对应 iOS 的 GoEngine.shared
 * 
 * 注意：这是一个简化版本，专注于实际使用场景
 */
class GoEngine private constructor(private val context: Context) {
    
    companion object {
        @Volatile
        private var instance: GoEngine? = null
        
        /**
         * 初始化 Go 引擎单例
         * 必须在 Application onCreate 中调用
         */
        fun initialize(context: Context): GoEngine {
            return instance ?: synchronized(this) {
                val newInstance = GoEngine(context.applicationContext)
                instance = newInstance
                newInstance
            }
        }
        
        /**
         * 获取 Go 引擎实例
         * 必须先调用 initialize()
         */
        fun shared(): GoEngine {
            return instance ?: throw IllegalStateException(
                "GoEngine not initialized. Call GoEngine.initialize() in Application.onCreate()"
            )
        }

        /**
         * 同步初始化 libbox 环境（专为 VPN 服务启动设计）
         */
        fun setupLibboxSync(context: Context) {
            val baseDir = context.filesDir
            val workingDir = context.getExternalFilesDir(null) ?: baseDir
            val tempDir = context.cacheDir
            try {
                val setupOptions = libbox.SetupOptions()
                setupOptions.basePath = baseDir.path
                setupOptions.workingPath = workingDir.path
                setupOptions.tempPath = tempDir.path
                setupOptions.fixAndroidStack = true
                // 使用端口 0 让系统自动分配可用端口，避免静态端口冲突
                setupOptions.commandServerListenPort = 0
                setupOptions.commandServerSecret = "OpenMesh-Secret-2026"
                libbox.Libbox.setup(setupOptions)
                
                // 对齐 iOS: 重定向 stderr 到文件以便排查 Go 层崩溃
                val stderrLogPath = java.io.File(context.cacheDir, "stderr.log").absolutePath
                try {
                    libbox.Libbox.redirectStderr(stderrLogPath)
                } catch (e: Exception) {
                    Log.w("GoEngine", "Failed to redirect stderr: ${e.message}")
                }

                // 对齐 iOS: 设置内存限制保护 (256MB)
                try {
                    libbox.Libbox.setMemoryLimit(true)
                } catch (e: Exception) {
                    Log.w("GoEngine", "Failed to set memory limit: ${e.message}")
                }

                Log.i("GoEngine", "libbox setup completed (sync). stderr.log: $stderrLogPath")
            } catch (e: Exception) {
                Log.e("GoEngine", "libbox setup failed (sync): ${e.message}")
            }
        }
    }
    
    private val TAG = "GoEngine"
    private val initialized = AtomicBoolean(false)
    
    @Volatile
    private var omLib: openmesh.AppLib? = null
    
    @Volatile
    private var appLib: OpenmeshAppLibProtocol? = null
    
    private var cachedConfig: ByteArray = ByteArray(0)
    
    /**
     * 懒初始化：第一次使用时才创建
     */
    init {
        Log.d(TAG, "GoEngine instance created with context: ${context.applicationContext?.javaClass?.simpleName}")
    }
    
    /**
     * 生成 12 个单词的助记词
     */
    suspend fun generateMnemonic12(): String = withContext(Dispatchers.IO) {
        ensureReady()
        appLib?.generateMnemonic12() ?: throw GoEngineError.NotInitialized
    }
    
    /**
     * 创建 EVM 钱包
     * @param mnemonic 助记词
     * @param pin PIN 码（用于加密 keystore）
     * @return 加密后的 keystore JSON 字符串
     */
    suspend fun createEvmWallet(mnemonic: String, pin: String): String = withContext(Dispatchers.IO) {
        ensureReady()
        appLib?.createEvmWallet(mnemonic, pin) ?: throw GoEngineError.NotInitialized
    }
    
    /**
     * 解密 EVM 钱包
     * @param keystoreJSON 加密的 keystore JSON
     * @param pin PIN 码
     * @return 解密后的钱包密钥信息
     */
    suspend fun decryptEvmWallet(keystoreJSON: String, pin: String): WalletSecretsV1 = withContext(Dispatchers.IO) {
        ensureReady()
        appLib?.decryptEvmWallet(keystoreJSON, pin) ?: throw GoEngineError.NotInitialized
    }
    
    /**
     * 获取代币余额
     */
    suspend fun getTokenBalance(address: String, tokenName: String, networkName: String): String = withContext(Dispatchers.IO) {
        val requestId = java.util.UUID.randomUUID().toString().take(8)
        val maskedAddr = maskAddress(address)
        Log.d(TAG, "getTokenBalance start request_id=$requestId address=$maskedAddr token=$tokenName network=$networkName")
        
        try {
            ensureReady()
            val balance = appLib?.getTokenBalance(address, tokenName, networkName) 
                ?: throw GoEngineError.NotInitialized
            Log.d(TAG, "getTokenBalance success request_id=$requestId balance=$balance")
            balance
        } catch (e: Exception) {
            Log.e(TAG, "getTokenBalance failed request_id=$requestId error=${e.message}")
            throw e
        }
    }
    
    /**
     * 获取支持的网络列表
     */
    suspend fun getSupportedNetworks(): List<String> = withContext(Dispatchers.IO) {
        ensureReady()
        val networksJson = appLib?.getSupportedNetworks() ?: throw GoEngineError.NotInitialized
        
        // 解析 JSON 数组
        try {
            org.json.JSONArray(networksJson).run {
                List(length()) { i -> getString(i) }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse networks JSON: $networksJson", e)
            throw e
        }
    }
    
    /**
     * 初始化或重新配置 Go 库
     * @param config 配置数据（JSON 格式）
     */
    suspend fun reconfigure(config: ByteArray) = withContext(Dispatchers.IO) {
        cachedConfig = config.copyOf()
        initializeLib(config)
    }
    
    /**
     * 重置 Go 引擎状态
     */
    suspend fun reset() = withContext(Dispatchers.IO) {
        omLib = null
        appLib = null
        cachedConfig = ByteArray(0)
        initialized.set(false)
        Log.d(TAG, "GoEngine reset completed")
    }
    
    /**
     * 释放运行时资源以响应内存压力
     * 用于 Android 退到后台后的内存回收
     */
    suspend fun releaseRuntimeForMemoryPressure() = withContext(Dispatchers.IO) {
        Log.d(TAG, "releaseRuntimeForMemoryPressure begin")
        omLib = null
        appLib = null
        Log.d(TAG, "releaseRuntimeForMemoryPressure end")
    }
    
    /**
     * 获取 VPN 状态
     */
    suspend fun getVpnStatus(): VpnStatus? = withContext(Dispatchers.IO) {
        try {
            Log.d(TAG, "getVpnStatus start")
            ensureReady()
            val status = appLib?.getVpnStatus()
            if (status != null) {
                Log.d(TAG, "getVpnStatus success connected=${status.connected} server=${status.server}")
            } else {
                Log.d(TAG, "getVpnStatus success nil")
            }
            status
        } catch (e: Exception) {
            Log.e(TAG, "getVpnStatus failed error=${e.message}")
            null
        }
    }
    
    // MARK: - Private methods
    
    /**
     * 确保 Go 库已就绪
     */
    private suspend fun ensureReady() {
        if (!initialized.get()) {
            initializeLib(cachedConfig)
        }
    }
    
    /**
     * 初始化 Go 库
     */
    private suspend fun initializeLib(config: ByteArray) = withContext(Dispatchers.IO) {
        try {
            if (omLib == null) {
                // 设置 gomobile 上下文
                go.Seq.setContext(context)
                
                // 初始化 libbox 环境
                val baseDir = context.filesDir
                val workingDir = context.getExternalFilesDir(null) ?: baseDir
                val tempDir = context.cacheDir
                try {
                    setupLibboxSync(context)
                } catch (e: Exception) {
                    Log.e(TAG, "libbox setup failed: ${e.message}")
                }

                // 使用 AAR 中 gomobile 生成的真实工厂方法
                val goLib = openmesh.Openmesh.newLib()
                    ?: throw GoEngineError.NewLibReturnedNil
                
                omLib = goLib
                appLib = OpenmeshAppLibBridge(goLib)
                Log.d(TAG, "OpenMeshGo library initialized")
            }
            
            if (config.isNotEmpty()) {
                appLib?.initApp(config)
            }
            
            initialized.set(true)
            Log.d(TAG, "GoEngine initialization completed")
        } catch (e: Exception) {
            Log.e(TAG, "GoEngine initialization failed: ${e.message}", e)
            throw e
        }
    }
    
    /**
     * 掩码钱包地址用于日志输出
     */
    private fun maskAddress(address: String): String {
        val trimmed = address.trim()
        if (trimmed.length <= 12) return trimmed
        return "${trimmed.take(6)}...${trimmed.takeLast(4)}"
    }
}
